[Gradient Compression] Add unit tests that test default Python comm hook implementations

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47158 [Gradient Compression] Add unit tests that test default Python comm hook implementations**

1. Test the default Python comm hook implementations ALLREDUCE and FP16_COMPRESS, besides an ad-hoc all-reduce implementation.
2. Typo fix.
3. Reformat default_hooks.py.
4. Publish register_comm_hook API for DDP module (This should be done in a separate diff, but got merged unintentionally.)

The new style can be used for testing any new comm hook like PowerSGD easily.
Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24669639](https://our.internmc.facebook.com/intern/diff/D24669639/)

Commit:
[Gradient Compression] Add unit tests that test default Python implementations

1. Test the default Python comm hook implementations ALLREDUCE and FP16_COMPRESS, besides an ad-hoc all-reduce implementation.
2. Typo fix.
3. Reformat default_hooks.py.

Differential Revision: [D24669639](https://our.internmc.facebook.com/intern/diff/D24669639/)

[ghstack-poisoned]

pritamdamania87 commented:
Looks like this test failure is relevant:

```
Oct 31 19:42:24 Traceback (most recent call last):
Oct 31 19:42:24   File "distributed/test_c10d.py", line 27, in <module>
Oct 31 19:42:24     import torch.distributed.algorithms.ddp_comm_hooks.default_hooks as default
Oct 31 19:42:24 ModuleNotFoundError: No module named 'torch.distributed.algorithms'
Oct 31 19:42:24 Traceback (most recent call last):
Oct 31 19:42:24   File "test/run_test.py", line 867, in <module>
Oct 31 19:42:24     main()
Oct 31 19:42:24   File "test/run_test.py", line 850, in main
Oct 31 19:42:24     raise RuntimeError(err_message)
Oct 31 19:42:24 RuntimeError: distributed/test_c10d failed!
```

Commit:
Update on "[Gradient Compression] Add unit tests that test default Python comm hook implementations"


1. Test the default Python comm hook implementations ALLREDUCE and FP16_COMPRESS, besides an ad-hoc all-reduce implementation.
2. Typo fix.
3. Reformat default_hooks.py.

The new style can be used for testing any new comm hook like PowerSGD easily.
Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24669639](https://our.internmc.facebook.com/intern/diff/D24669639/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add unit tests that test default Python comm hook implementations"


1. Test the default Python comm hook implementations ALLREDUCE and FP16_COMPRESS, besides an ad-hoc all-reduce implementation.
2. Typo fix.
3. Reformat default_hooks.py.

The new style can be used for testing any new comm hook like PowerSGD easily.
Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24669639](https://our.internmc.facebook.com/intern/diff/D24669639/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add unit tests that test default Python comm hook implementations"


1. Test the default Python comm hook implementations ALLREDUCE and FP16_COMPRESS, besides an ad-hoc all-reduce implementation.
2. Typo fix.
3. Reformat default_hooks.py.

The new style can be used for testing any new comm hook like PowerSGD easily.
Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24669639](https://our.internmc.facebook.com/intern/diff/D24669639/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add unit tests that test default Python comm hook implementations"


1. Test the default Python comm hook implementations ALLREDUCE and FP16_COMPRESS, besides an ad-hoc all-reduce implementation.
2. Typo fix.
3. Reformat default_hooks.py.

The new style can be used for testing any new comm hook like PowerSGD easily.
Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24669639](https://our.internmc.facebook.com/intern/diff/D24669639/)

[ghstack-poisoned]

SciPioneer commented:
> Looks like this test failure is relevant:
> 
> ```
> Oct 31 19:42:24 Traceback (most recent call last):
> Oct 31 19:42:24   File "distributed/test_c10d.py", line 27, in <module>
> Oct 31 19:42:24     import torch.distributed.algorithms.ddp_comm_hooks.default_hooks as default
> Oct 31 19:42:24 ModuleNotFoundError: No module named 'torch.distributed.algorithms'
> Oct 31 19:42:24 Traceback (most recent call last):
> Oct 31 19:42:24   File "test/run_test.py", line 867, in <module>
> Oct 31 19:42:24     main()
> Oct 31 19:42:24   File "test/run_test.py", line 850, in main
> Oct 31 19:42:24     raise RuntimeError(err_message)
> Oct 31 19:42:24 RuntimeError: distributed/test_c10d failed!
> ```
Thanks for catching this! 

Fixed by importing the relative, not absolute path of default hooks in torch/distributed/algorithms/ddp_comm_hooks/__init__.py

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -3523,6 +3551,12 @@ def test_builtin_ddp_comm_hooks_nccl(self):
+    @skip_if_lt_x_gpu(2)
+    @skip_if_rocm
+    def test_default_ddp_comm_hooks_nccl_is_view(self):
+        self._test_default_ddp_comm_hooks_nccl(gradient_as_bucket_view=True)
```
Thanks for adding the test with `gradient_as_bucket_view`! LGTM once all tests pass

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:
```diff
@@ -31,17 +31,19 @@ def then_callback(fut):
-
-        Example::
-            >>> ddp_model._register_comm_hook(process_group, fp16_compress_hook)
+    This DDP communication hook implements a simple gradient compression
```
These changes are fine because it's a small diff, but in general can we try to keep formatting changes in a separate diff? It makes the actual changes a lot easier to review. Thanks!

SciPioneer replied:
Sure. I also prefer separating out these noises. The formatting was done unintentionally. Thanks for the reminder!

Commit:
Update on "[Gradient Compression] Add unit tests that test default Python comm hook implementations"


1. Test the default Python comm hook implementations ALLREDUCE and FP16_COMPRESS, besides an ad-hoc all-reduce implementation.
2. Typo fix.
3. Reformat default_hooks.py.

The new style can be used for testing any new comm hook like PowerSGD easily.
Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24669639](https://our.internmc.facebook.com/intern/diff/D24669639/)

[ghstack-poisoned]
